packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,342 @@
+}
+
+// We only type the properties requested in the GraphQL query. This extends SnapshotProposalIpfs, but we need to
+// override the space property.
+interface SnapshotProposalGraphql extends Omit<SnapshotProposalIpfs, "space"> {
+  state: string;
+  space: { id: string };
+  scores: number[];
+  quorum: number;
+  scores_total: number;
+}
+
+interface GraphqlData {
+  proposals: SnapshotProposalGraphql[];
+}
+
+interface IpfsData {
+  data: {
+    message: SnapshotProposalIpfs;
+  };
+}
+
+export interface VerificationResponse {
+  verified: boolean;
+  error?: string;
+}
+
+interface RulesParameters {
+  space: string;
+  quorum: number;
+  votingPeriod: number;
+}
+
+// Returns null if the rules string does not match the expected template.
+const parseRules = (rules: string): RulesParameters | null => {
+  // This is based on the template from Zodiac app at
+  // https://github.com/gnosis/zodiac-safe-app/blob/79dbb72af506f60fcc16599516ce48f893393b29/packages/app/src/views/AddModule/wizards/OptimisticGovernorModule/OptimisticGovernorModuleModal.tsx#L136
+  const regex = /^I assert that this transaction proposal is valid according to the following rules: Proposals approved on Snapshot, as verified at https:\/\/snapshot\.org\/#\/(.*), are valid as long as there is a minimum quorum of (\d+) and a minimum voting period of (\d+) hours and it does not appear that the Snapshot voting system is being exploited or is otherwise unavailable\. The quorum and voting period are minimum requirements for a proposal to be valid\. Quorum and voting period values set for a specific proposal in Snapshot should be used if they are more strict than the rules parameter\. The explanation included with the on-chain proposal must be the unique IPFS identifier for the specific Snapshot proposal that was approved or a unique identifier for a proposal in an alternative voting system approved by DAO social consensus if Snapshot is being exploited or is otherwise unavailable.$/;
+
+  const match = rules.match(regex);
+  if (!match) {
+    return null;
+  }
+
+  const space = match[1];
+  const quorum = parseInt(match[2]);
+  const votingPeriod = parseInt(match[3]);
+
+  return { space, quorum, votingPeriod };
+};
+
+// We don't want to throw an error if the GraphQL request fails for any reason, so we return a stringified Error object
+// instead that will be logged by the bot.
+const getGraphqlData = async (ipfsHash: string, url: string): Promise<GraphqlData | Error> => {
+  const query = gql(/* GraphQL */ `
+    query GetProposals($ipfsHash: String) {
+      proposals(first: 2, where: { ipfs: $ipfsHash }, orderBy: "created", orderDirection: desc) {
+        type
+        choices
+        start
+        end
+        state
+        space {
+          id
+        }
+        scores
+        quorum
+        scores_total
+        plugins
+      }
+    }
+  `);
+  try {
+    return await request<GraphqlData, { ipfsHash: string }>(url, query, { ipfsHash });
```

Comment:
Are retries baked into this library?

Reply:
No, but I should fix this in a separate PR